src/inspector_profiler.cc

```diff
@@ -503,6 +507,246 @@ static void StopCoverage(const FunctionCallbackInfo<Value>& args) {
+  }
+}
+
+void ProfileWatchdog::Start(const FunctionCallbackInfo<Value>& args) {
+  Environment* env = Environment::GetCurrent(args);
+  ProfileWatchdog* watchdog = Unwrap<ProfileWatchdog>(args.Holder());
+  if (watchdog) {
+    watchdog->Start(env);
+  }
+}
+
+void ProfileWatchdog::Stop(const FunctionCallbackInfo<Value>& args) {
+  ProfileWatchdog* watchdog = Unwrap<ProfileWatchdog>(args.Holder());
+  if (watchdog) {
+    watchdog->Stop();
+  }
+}
+
+void ProfileWatchdog::Run(void* arg) {
+  ProfileWatchdog* wd = static_cast<ProfileWatchdog*>(arg);
+  uv_run(&wd->loop_, UV_RUN_DEFAULT);
+  CheckedUvLoopClose(&wd->loop_);
+}
+
+void ProfileWatchdog::Timer(uv_timer_t* timer) {
+  ProfileWatchdog* w = ContainerOf(&ProfileWatchdog::timer_, timer);
+  Environment* env = w->env();
+  if (!env || env->is_stopping()) {
+    return;
+  }
+  if (w->state() == ProfileWatchdogState::kClosed) {
+    return;
+  }
+  env->RequestInterrupt([watchdog = std::move(w)](Environment* env) {
+    if (env->is_stopping() ||
+        watchdog->state() == ProfileWatchdogState::kClosed) {
+      return;
+    }
+    if (watchdog->state() == ProfileWatchdogState::kClosing) {
+      delete watchdog;
+      return;
+    }
+    if (watchdog->TimeoutHandler()) {
+      watchdog->AddTask(
+          [watchdog = std::move(watchdog)]() { watchdog->SetTimeout(); });
+    }
+  });
+}
+
+template <typename Fn>
+void ProfileWatchdog::AddTask(Fn&& cb, CallbackFlags::Flags flags) {
+  auto callback = tasks_.CreateCallback(std::move(cb), flags);
+  {
+    Mutex::ScopedLock lock(task_mutex_);
+    tasks_.Push(std::move(callback));
+  }
+  uv_async_send(&async_);
+}
+
+void ProfileWatchdog::HandleTasks() {
+  while (tasks_.size() > 0) {
+    CallbackQueue<void> queue;
+    {
+      Mutex::ScopedLock lock(task_mutex_);
+      queue.ConcatMove(std::move(tasks_));
+    }
+    while (auto head = queue.Shift()) head->Call();
+  }
+}
+
+void ProfileWatchdog::SetTimeout() {
+  uv_timer_start(&timer_, &ProfileWatchdog::Timer, interval_, 0);
+}
+
+Local<FunctionTemplate> ProfileWatchdog::GetConstructorTemplate(
+    Environment* env) {
+  Local<FunctionTemplate> tmpl = env->profile_watchdog_ctor_template();
+  if (tmpl.IsEmpty()) {
+    Isolate* isolate = env->isolate();
+    tmpl = NewFunctionTemplate(isolate, nullptr);
+    tmpl->SetClassName(
+        FIXED_ONE_BYTE_STRING(env->isolate(), "ProfileWatchdog"));
+    SetProtoMethod(isolate, tmpl, "start", ProfileWatchdog::Start);
+    SetProtoMethod(isolate, tmpl, "stop", ProfileWatchdog::Stop);
+    env->set_profile_watchdog_ctor_template(tmpl);
+  }
+  return tmpl;
+}
+
+ProfileWatchdog::ProfileWatchdog(Environment* env, Local<Object> object)
+    : BaseObject(env, object) {}
+
+ProfileWatchdog::~ProfileWatchdog() {
+  ProfileWatchdogState old_state = state();
+  set_state(ProfileWatchdogState::kClosed);
+  if (old_state != ProfileWatchdogState::kInitialized) {
+    AddTask([self = std::move(this)]() {
```

Comment:
This looks like a use-after-free. I suspect you were thinking of `std::move(*this)` (note the star) but even that isn't correct.